commonmark/src/main/java/org/commonmark/internal/DocumentParser.java

```diff
@@ -81,6 +84,27 @@ public Document parse(String input) {
         this.processInlines();
         return documentBlockParser.getBlock();
     }
+    
+    public Document parse(Reader input) throws IOException {
+        DocumentBlockParser documentBlockParser = new DocumentBlockParser();
+        activateBlockParser(documentBlockParser);
+
+        BufferedReader bufferedReader;
+        if (input instanceof BufferedReader) {
+            bufferedReader = (BufferedReader) input;
+        } else {
+            bufferedReader = new BufferedReader(input);
+        }
+        
+        String line;
+        while ((line = bufferedReader.readLine()) != null) {
+            incorporateLine(line);
+        }
+
+        finalizeBlocks(activeBlockParsers);
+        this.processInlines();
+        return documentBlockParser.getBlock();
```

Comment:
These three lines could also be shared by extracting them into a method and reusing it.